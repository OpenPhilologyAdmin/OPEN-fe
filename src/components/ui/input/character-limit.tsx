import { ComponentPropsWithoutRef, useEffect, useState } from "react";

import styled from "styled-components";

import Typography from "../../ui/typography";

type UseCharacterLimitProps = string | number | readonly string[] | undefined;
type CharacterLimitProps = ComponentPropsWithoutRef<"span"> & {
  current: string | number | undefined;
  max: string | number;
};

const Label = styled(Typography)`
  display: flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  padding: 0 8px;
  height: 20px;
  border-radius: ${({ theme: { borderRadius } }) => `${borderRadius.sm} 0 ${borderRadius.sm} 0`};
  background-color: ${({ theme: { colors } }) => colors.backgroundSecondary};
  mix-blend-mode: multiply;
  color: ${({ theme: { colors } }) => colors.textTertiary};
`;

export function useCharacterLimit(value: UseCharacterLimitProps) {
  const [current, setCurrent] = useState(String(value || "").length);

  useEffect(() => {
    setCurrent(String(value || "").length);
  }, [value]);

  return { current };
}

function CharacterLimit({ current, max, ...props }: CharacterLimitProps) {
  if (Number(max) === 0) return null;

  return (
    <Label variant="tiny" {...props}>
      {current}/{max}
    </Label>
  );
}

export default CharacterLimit;
