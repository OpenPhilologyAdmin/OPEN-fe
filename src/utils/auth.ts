import { COOKIES } from "@/constants/cookies";
import clientCookie from "js-cookie";

export const getAccessToken = (): string | undefined => {
  return clientCookie.get(COOKIES.ACCESS_TOKEN);
};

export const setAccessToken = (token: string) => {
  clientCookie.set(COOKIES.ACCESS_TOKEN, token);
};

export const removeAccessToken = () => {
  clientCookie.remove(COOKIES.ACCESS_TOKEN);
};
