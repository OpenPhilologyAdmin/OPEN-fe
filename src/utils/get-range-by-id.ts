export const getRangeById = <ElementType extends { id: number }>(
  elements: ElementType[],
  startElementId: number,
  endElementId: number,
) => {
  const elementIds = elements.map(element => element.id);
  const startElementIndex = elementIds.indexOf(startElementId);
  const endElementIndex = elementIds.indexOf(endElementId);

  // only one element followed by apparatus index was selected
  if (endElementId === Infinity) {
    return [elements[startElementIndex]];
  }

  if (startElementId === Infinity) {
    return [elements[endElementIndex]];
  }

  return startElementIndex < endElementIndex
    ? elements.slice(startElementIndex, endElementIndex + 1)
    : elements.slice(endElementIndex, startElementIndex + 1);
};
