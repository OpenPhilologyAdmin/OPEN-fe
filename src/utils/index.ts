// placeholder file for structure
export const placeholder = "";
