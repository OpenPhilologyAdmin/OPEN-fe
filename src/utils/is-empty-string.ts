export const isEmptyString = (str?: string) => str === "";
