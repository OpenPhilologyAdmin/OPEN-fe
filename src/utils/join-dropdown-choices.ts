import { NormalizedDropdownOption } from "@/components/dropdown/item";

export function joinDropdownChoices(options: NormalizedDropdownOption[], prompt: string): string {
  const selectedOptions = options.filter(option => Boolean(option.selected));

  if (!(selectedOptions.length > 0)) return prompt;

  return selectedOptions.map(option => option.label).join(", ");
}
