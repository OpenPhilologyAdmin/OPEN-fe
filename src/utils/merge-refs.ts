export function mergeRefs<GenericElement extends HTMLElement | null = HTMLDivElement>(
  ...refs: (React.MutableRefObject<GenericElement> | React.Ref<GenericElement> | undefined)[]
) {
  return (value: GenericElement) => {
    refs.forEach(ref => {
      if (typeof ref === "function") {
        ref(value);
      } else if (ref) {
        (ref as React.MutableRefObject<GenericElement>).current = value;
      }
    });
  };
}
