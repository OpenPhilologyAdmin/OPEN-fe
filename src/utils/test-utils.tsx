import { ReactNode } from "react";
import ModalMock from "react-modal";
import { QueryClientProvider } from "react-query";
import { RouterContext } from "next/dist/shared/lib/router-context";
import { NextRouter } from "next/router";

import { ToastProvider } from "@/components/ui/toast";
import { UserProvider } from "@/contexts/user";
import { mswServer } from "@/mocks/index";
import { mockQueryCache, mockQueryClient } from "@/mocks/mock-query";
import { mockRouter } from "@/mocks/mock-router";
import { mockUser } from "@/mocks/mock-user";
import * as testing from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeProvider } from "src/contexts/theme";

beforeAll(() => {
  mockQueryCache.clear();
  mswServer.listen({
    onUnhandledRequest: "error",
  });
});

afterEach(() => {
  mswServer.resetHandlers();
  mockQueryCache.clear();
});

afterAll(() => mswServer.close());

function MockProvider({ children, user }: { children: ReactNode; user?: API.User }) {
  return (
    <ThemeProvider initialTheme="LIGHT">
      <UserProvider initialUser={user}>
        <QueryClientProvider client={mockQueryClient}>{children}</QueryClientProvider>
      </UserProvider>
    </ThemeProvider>
  );
}

/**
 * Include this provider in tests that rely on react-tostify toast component
 *
 * This is not mocked globally because the way the library is written, it would disrupt all existing snapshots by injecting the provider
 */
export function MockToastProvider() {
  return <ToastProvider />;
}

type DefaultParams = Parameters<typeof testing.render>;
type RenderUI = DefaultParams[0];
type RenderOptions = Omit<DefaultParams[1], "queries"> & {
  router?: Partial<NextRouter>;
  user?: Partial<API.User>;
};

/**
 * Custom render function that wraps the default render from @testing-library/react
 *
 * This function is used to wrap components rendered in tests with mandatory providers used in the app
 *
 * @param ui The rendered component
 * @param options Default render options extended by mocked NextRouter
 * @example render(<Component />, { router: { query: { token: 123 } } })
 */
const customRender = (ui: RenderUI, options?: RenderOptions) => {
  const wrapper = ({ children }: { children: ReactNode }) => (
    <MockProvider user={{ ...mockUser, ...options?.user }}>
      {options?.router ? (
        <RouterContext.Provider value={{ ...mockRouter, ...options.router }}>
          {children}
        </RouterContext.Provider>
      ) : (
        children
      )}
    </MockProvider>
  );

  return testing.render(ui, { wrapper, ...options });
};

// Mock external modal library
const oldFn = ModalMock.setAppElement;

ModalMock.setAppElement = element => {
  if (element === "#__next") {
    // otherwise it will throw aria warnings.
    return oldFn(document.createElement("div"));
  }

  oldFn(element);
};

// ! We want to export all from testing library and override custom render on purpose
// eslint-disable-next-line import/export
export * from "@testing-library/react";
// eslint-disable-next-line import/export
export { customRender as render, mswServer as mockServer, userEvent };
