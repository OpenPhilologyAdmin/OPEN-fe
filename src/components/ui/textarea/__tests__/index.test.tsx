import { render, renderHook, screen, userEvent } from "@/utils/test-utils";

import { useCharacterLimit } from "../../character-limit";
import TextArea from "..";

describe("TextArea", () => {
  it("renders a TextArea unchanged", () => {
    const { container } = render(<TextArea />);

    expect(container).toMatchSnapshot();
  });

  it("renders a TextArea with label", () => {
    const { label, id } = { label: "label", id: "id" };

    render(<TextArea id={id} label={label} />);

    expect(screen.getByLabelText(label));
  });

  it("renders a TextArea with CharacterLimit", () => {
    const { defaultValue, ...textAreaProps } = {
      maxLength: 10,
      defaultValue: "asd",
    };

    const { result: characterLimit } = renderHook(() => {
      return useCharacterLimit(defaultValue);
    });

    render(<TextArea current={characterLimit.current.current} {...textAreaProps} />);

    expect(screen.getByText(`${defaultValue.length}/${textAreaProps.maxLength}`));
  });

  it("renders a TextArea with ErrorLabel", () => {
    const errorMessage = "errorMessage";

    render(<TextArea invalid errorMessage={errorMessage} />);

    expect(screen.getByText(errorMessage));
  });

  it("renders a TextArea without ErrorLabel when valid", () => {
    const errorMessage = "errorMessage";

    render(<TextArea invalid={false} errorMessage={errorMessage} />);

    expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
  });

  it("renders a TextArea and focuses on keyboard navigation", async () => {
    const { label, id } = { label: "label", id: "id" };
    const user = userEvent.setup();

    render(<TextArea id={id} label={label} />);

    const textArea = screen.getByLabelText(label);

    await user.keyboard("[Tab]");

    expect(document.activeElement).toBe(textArea);
  });

  it("renders a TextArea and focuses on click", async () => {
    const { label, id } = { label: "label", id: "id" };
    const user = userEvent.setup();

    render(<TextArea id={id} label={label} />);

    const textArea = screen.getByLabelText(label);

    await user.click(textArea);

    expect(document.activeElement).toBe(textArea);
  });
});
