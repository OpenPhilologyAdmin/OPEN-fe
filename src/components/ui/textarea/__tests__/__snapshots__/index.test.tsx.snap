// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders a TextArea unchanged 1`] = `
.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 48px;
  width: 100%;
  border: solid 2px #868686;
  border-radius: 4px;
  background-color: #FFFFFF;
  cursor: text;
}

.c1:focus-within {
  box-shadow: 0 0 0 4px rgba(93,45,239,0.32);
}

.c1::after {
  content: "";
  position: absolute;
  display: none;
  position: absolute;
  z-index: 2;
  bottom: 4px;
  right: 4px;
  width: 12px;
  height: 12px;
  background-image: url("/images/icons/textarea-resize-custom-icon.svg");
  background-repeat: no-repeat;
  pointer-events: none;
}

.c2 {
  position: relative;
  z-index: 1;
  outline: none;
  padding: 12px;
  opacity: 1;
  margin: 0;
  min-width: 0;
  width: 100%;
  border: none;
  min-height: 48px;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  border-radius: 4px;
  color: #101521;
  resize: none;
  background-color: transparent;
  cursor: text;
}

.c2::-webkit-input-placeholder {
  color: #101521;
}

.c2::-moz-placeholder {
  color: #101521;
}

.c2:-ms-input-placeholder {
  color: #101521;
}

.c2::placeholder {
  color: #101521;
}

.c2::-webkit-resizer {
  display: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 100%;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <textarea
        class="c2"
      />
    </div>
  </div>
</div>
`;
